libcxx/include/string

```diff
@@ -1891,8 +1891,7 @@ private:
 #if !defined(_LIBCPP_HAS_NO_ASAN) && defined(_LIBCPP_INSTRUMENTED_WITH_ASAN)
     const void* __begin = data();
     const void* __end   = data() + capacity() + 1;
-    if (!__libcpp_is_constant_evaluated() && __begin != nullptr &&
-        is_same<allocator_type, __default_allocator_type>::value)
+    if (!__libcpp_is_constant_evaluated() && __asan_annotate_container_with_allocator<allocator_type>::value)
```

Comment:
In your example, the string is short (btw. the default allocator is also used), and until https://github.com/llvm/llvm-project/pull/75882 annotations for SSO cases are turned off. But even if you turn them on (change value of `_LIBCPP_SHORT_STRING_ANNOTATIONS_ALLOWED` to true) it works as expected. Constant evaluated objects are not annotated, runtime objects are annotated. s2 is annotated when it's created.

s is probably not annotated, but it's ok, it may only lead to false negatives. I will investigate it before merge of short string annotations, but it shouldn't affect this PR.

No boom is expected. No boom happens.

Reply:
It can tolerate unpoisoned memory where it should be poisoned, not the other way around. It may result in some false negatives, but we accept that.